stdlib/bz2.pyi

```diff
@@ -80,6 +80,15 @@ def open(
     errors: str | None = ...,
     newline: str | None = ...,
 ) -> TextIO: ...
+@overload
+def open(
+    filename: StrOrBytesPath | _ReadableFileobj | _WritableFileobj,
+    mode: str,
+    compresslevel: int = ...,
+    encoding: str | None = ...,
+    errors: str | None = ...,
+    newline: str | None = ...,
+) -> BZ2File | TextIO: ...
```

Comment:
A union return type would be annoying in practice. This won't work:

```
def func(mode: str) -> TextIO:
    return bz2.open("some path", mode)
```

because the type checker will warn that the return value could also be a `BZ2File`, which supports `IO[bytes]`, not `TextIO`.

```suggestion
) -> IO[Any]: ...
```

Reply:
This is how `lzma.open` and `gzip.open` do it as well. So changing this would add an inconsistency.

If this change is desired it should be changed in `lzma.open` and `gzip.open` as well. But it is maybe best to not do all that in this PR.

Reply:
Thinking a bit more about it, while we usually avoid union return types, in this case I'm fine with the more precise type. Most cases should be covered by the literal overloads, this case only comes into play when an unknown mode is used. It stands to reason that in this case the user needs to ensure that the returned objects is of the correct type before using it.